### PR TITLE
Fix potential use of NULL pointers

### PR DESCRIPTION
**Pull request by cgohlke**
## Description

Always verify that malloc succeeded.

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by hmaarrfk** [comment]
C++ got alot better when they standardized smart pointers.

**Comment by stefanv** [comment]
What can we do to safeguard ourselves from all the various types of memory mistakes that we've been seeing?

File: skimage/measure/_marching_cubes_lewiner_cy.pyx [hunk 1]
```diff
@@ -190,6 +187,13 @@ cdef class Cell:
         # Allocate face layers
         self.faceLayer1 = <int *>malloc(self.nx*self.ny*4 * sizeof(int))
         self.faceLayer2 = <int *>malloc(self.nx*self.ny*4 * sizeof(int))
+
+        if (self.faceLayer1 is NULL or self.faceLayer2 is NULL or
+            self.vv is NULL or self.vg is NULL or self._vertices is NULL or
+            self._normals is NULL or self._values is NULL or
+            self._faces is NULL):
+            raise MemoryError()
+
```
Comment (stefanv) [inline skimage/measure/_marching_cubes_lewiner_cy.pyx]: What happens here if we do not de-allocate the values that already succeeded?

**Comment by hmaarrfk** [comment]
you can just use numpy to allocate arrays. there is no reason to use malloc today. Maybe in the old days of cython that wasn't so true.

**Comment by hmaarrfk** [comment]
You can still call numpy from within cython, it is just rather annoying. It is done in a few locations in skimage already.

**Comment by hmaarrfk** [comment]
Yeah, but it is harder to "review" and makes the cython functions really verbose.

**Comment by cgohlke** [comment]
> you can just use numpy to allocate arrays.

can't do that without GIL. Not sure if that'll work for OpenMP thread local buffers.

File: skimage/measure/_marching_cubes_lewiner_cy.pyx (hunk 1 above)
Comment (cgohlke) [reply skimage/measure/_marching_cubes_lewiner_cy.pyx]: The should be released in `__dealloc__`, no?

Comment (hmaarrfk) [reply skimage/measure/_marching_cubes_lewiner_cy.pyx]: is __dealloc__ safe to any failed memory allocations? seems like they all need to be guarded right?

Comment (jni) [reply skimage/measure/_marching_cubes_lewiner_cy.pyx]: Yes, `free` is a no-op with NULL pointers, and will work correctly with allocated pointers. It is *not* safe to `free` a pointer that has already been freed. However that should not happen in this case, I believe.

**Pull request merged**